core/src/main/java/com/oath/oak/ValueUtils.java

```diff
@@ -1,368 +1,58 @@
-            oldHeader = bb.getInt(bb.position());
-            if (isValueDeleted(oldHeader)) {
-                return FAILURE;
-            }
-            if (wasValueMoved(oldHeader)) {
-                return ValueResult.MOVED;
-            }
-            oldHeader &= ~LOCK_MASK;
-        } while (!CAS(bb, oldHeader, oldHeader + 4));
-        return SUCCESS;
-    }
+    ByteBuffer getValueByteBufferNoHeaderPrivate(Slice s);
 
-    /**
-     * @see #unlockRead(ByteBuffer)
-     */
-    static void unlockRead(Slice s) {
-        unlockRead(s.getByteBuffer());
-    }
+    ByteBuffer getValueByteBufferNoHeader(Slice s);
 
-    /**
-     * Releases a read lock of the given value.
-     *
-     * @param bb the value Byte Buffer (including the header)
-     */
-    static void unlockRead(ByteBuffer bb) {
-        assert bb.isDirect();
-        int oldHeader;
-        do {
-            oldHeader = bb.getInt(bb.position());
-        } while (!CAS(bb, oldHeader, oldHeader - 4));
-    }
+    ValueResult lockRead(Slice s, int version);
 
-    private static ValueResult lockWrite(ByteBuffer bb) {
-        assert bb.isDirect();
-        int oldHeader;
-        do {
-            oldHeader = bb.getInt(bb.position());
-            if (isValueDeleted(oldHeader)) {
-                return FAILURE;
-            }
-            if (wasValueMoved(oldHeader)) {
-                return ValueResult.MOVED;
-            }
-        } while (!CAS(bb, FREE.value, LOCKED.value));
-        return SUCCESS;
-    }
+    ValueResult unlockRead(Slice s, int version);
 
-    private static void unlockWrite(Slice s) {
-        unlockWrite(s.getByteBuffer());
-    }
+    ValueResult lockWrite(Slice s, int version);
 
-    private static void unlockWrite(ByteBuffer bb) {
-        bb.putInt(bb.position(), FREE.value);
-        // maybe a fence?
-    }
+    ValueResult unlockWrite(Slice s);
 
-    /* Handle Methods */
+    ValueResult deleteValue(Slice s, int version);
 
-    static void unsafeBufferToIntArrayCopy(ByteBuffer bb, int srcPosition, int[] dstArray, int countInts) {
-        UnsafeUtils.unsafeCopyBufferToIntArray(bb, srcPosition, dstArray, countInts);
-    }
+    ValueResult isValueDeleted(Slice s, int version);
 
-    /**
-     * Used only in OakRValueBuffer
-     *
-     * @param bb          the value Byte Buffer (including the header)
-     * @param transformer value deserializer
-     * @param <T>         the type of {@code transformer}'s output
-     * @return {@code null} if the value is deleted or was moved.
-     * The value written off-heap is returned otherwise.
-     */
-    static <T> T transform(ByteBuffer bb, Function<ByteBuffer, T> transformer) {
-        if (lockRead(bb) != SUCCESS) {
-            return null;
-        }
+    int getOffHeapVersion(Slice s);
 
-        T transformation = transformer.apply(getValueByteBufferNoHeader(bb).asReadOnlyBuffer());
-        unlockRead(bb);
-        return transformation;
-    }
+    int NOVA_HEADER_SIZE = 4;
```

Comment:
Should it be named with "NOVA". ValueUtils should work regardless to memory management...

Reply:
This is the size of the version. It does not have to necessarily be connected to Nova, but it has to be connected to versions (as does ValueUtils). I renamed it to ``VERSION_SIZE``.